### PR TITLE
Long Strip Split for Webtoon

### PR DESCRIPTION
**Pull request by AntsyLich**
The splitting is on by default to know how it may effect different devices

Continues #4791 

Thanks @ghostbear for helping out time to time.

### CODE REVIEW

**Comment by Riztard** [comment]
> Removes Dual Page Split from Webtoon reader. Doesn't have any actual purpose and is not needed for webtoon reader.

some people use webtoon reader for normal manga tho

**Comment by FourTOne5** [comment]
It just inverts the image if dual page invert is on Incase of Dual Page. So
that's why I said it doesn't have any actual purpose.

**Comment by ghostbear** [comment]
If a dual page is detected they are actually stacked instead of displayed side by side

```
From:
*-----*-----*
|     |     |
|     |     |
|     |     |
*-----*-----*

To:
*-----*
|     |
|     |
|     |
*-----*
|     |
|     |
|     |
*-----*
```

Invert just changes where the pages are placed

This is useful for people who read old manga/shitty scans where all pages are doubled in webtoon mode

**Comment by FourTOne5** [comment]
> If a dual page is detected they are actually stacked instead of displayed side by side
> ¤ Only the case for Webtoon, the Pager implementation is different
> 
> ```
> From:
> *-----*-----*
> |     |     |
> |     |     |
> |     |     |
> *-----*-----*
> 
> To:
> *-----*
> |     |
> |     |
> |     |
> *-----*
> |     |
> |     |
> |     |
> *-----*
> ```
> 
> Invert just changes where the pages are placed
> 
> This is useful for people who read old manga/shitty scans where all pages are doubled in webtoon mode

Yeah I actually noticed it after writing the previous comment and rechecking it's function.

**Comment by FourTOne5** [comment]
Can someone test this? Android Studio is refusing to create a build for me (actually my PC starts lagging intensely 😐)

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/webtoon/WebtoonAdapter.kt [hunk 1]
```diff
@@ -5,10 +5,7 @@ import android.widget.FrameLayout
 import android.widget.LinearLayout
 import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.RecyclerView
-import eu.kanade.tachiyomi.ui.reader.model.ChapterTransition
-import eu.kanade.tachiyomi.ui.reader.model.ReaderChapter
-import eu.kanade.tachiyomi.ui.reader.model.ReaderPage
-import eu.kanade.tachiyomi.ui.reader.model.ViewerChapters
+import eu.kanade.tachiyomi.ui.reader.model.*
```
Comment (Riztard) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/webtoon/WebtoonAdapter.kt:8]: dont use wildcard?

Comment (FourTOne5) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/webtoon/WebtoonAdapter.kt:8]: Android Studio is automatically doing that. I think I'll have to disable it.

**Comment by FourTOne5** [comment]
Hmm... Everything is blank. 

Screenshot_20210823_192421

**Comment by FourTOne5** [comment]
@ghostbear can you check why the images are blank? I have zero idea on what I should do.

**Comment by ghostbear** [comment]
You have to reset the InputStream whenever you read from it either with `mark` and `reset` or by creating a new `ByteArrayInputStream`. If you don't reset the InputStream it will result in a blank image

https://stackoverflow.com/questions/9501237/read-input-stream-twice

**Comment by specterflare** [comment]
Not really sure if there's any way to "bump" this, but I just wanted to say that I had to ditch a webtoon because like 80 chapters in it switched from dividing chapters into 30+ pages to just 4 or 5, which causeds Tachiyomi to leep crashing, so any work you guys do with this is greatly appreciated!

**Comment by FourTOne5** [comment]
> Not really sure if there's any way to "bump" this, but I just wanted to say that I had to ditch a webtoon because like 80 chapters in it switched from dividing chapters into 30+ pages to just 4 or 5, which caused Tachiyomi to keep crashing, so any work you guys do with this is greatly appreciated!

I'm busy with very important exam or else I would have worked on this. I just need to address crash on splitting (which is the only problem I'm aware off).
```kt
fun stripSplitCount(imageStream: InputStream): Int {
        imageStream.mark(imageStream.available() + 1)

        val imageBytes = imageStream.readBytes()

        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size, options)

        imageStream.reset()
        val imageHeightForSplit = maxImageHeightForSplit(imageStream)

        return options.outHeight /  (imageHeightForSplit - 1)
}
```
Will probably fix it but I'm quite not sure.
@ghostbear due you mind taking over? I'll start working at end of November if you don't.

**Comment by YurDavid** [comment]
> m or else I would have worked on this. I just need to address crash on splitting (which is the only problem I'm aware off).



> Continues #4791 with suggestion from @ghostbear
> 
> Needs testing so keeping it as draft.



> Continues #4791 with suggestion from @ghostbear
> 
> Needs testing so keeping it as draft.

sigh, it was painful how the long strips images don't load to fit the original size, so though could come around see if there is anything I could help with it,

**Comment by iKlsR** [comment]
Eagerly awaiting this.

**Comment by FourTOne5** [comment]
> Eagerly awaiting this. Let me know if I can help in any way.

Yes by rewriting the reader so it can handle the strip splitting smoothly

**Comment by saud-97** [comment]
Hi there @FourTOne5 and all, really appreciate the work done. I noticed a message above "All checks have passed" and "This pull request is still a work in progress".

Does that mean the feature is complete now and just waiting for review or does it still some work?

**Comment by AntsyLich** [comment]
Conflicts has been resolved and the PR is ready to be reviewed.

File: app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt [hunk 2]
```diff
@@ -293,6 +297,65 @@ object ImageUtil {
...
+     */
+    fun isStripSplitNeeded(imageStream: BufferedInputStream): Boolean {
+        val options = extractImageOptions(imageStream)
+
+        val imageHeightIsBiggerThanWidth = options.outHeight > options.outWidth
+        val imageHeightBiggerThanScreenHeight = options.outHeight > getDisplayMaxHeightInPx
+        return imageHeightIsBiggerThanWidth && imageHeightBiggerThanScreenHeight
+    }
+
+    fun stripSplitData(imageStream: InputStream): List<SplitData> {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt:312]: A bunch of this seems repeated with `splitTallImage`. Should try to consolidate that so they don't drift.

File: app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt [hunk 3]
```diff
@@ -293,6 +297,65 @@ object ImageUtil {
         val bottomOffset = topOffset + outputImageHeight
     }
 
+    /**
+     * Check whether the image is a long Strip that needs splitting
+     * @return true if the image height is greater than image width and screen height
+     */
+    fun isStripSplitNeeded(imageStream: BufferedInputStream): Boolean {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt:304]: Should we be checking `isAnimatedAndSupported` here?

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -296,6 +296,8 @@
     <string name="pref_dual_page_split">Dual page split</string>
     <string name="pref_dual_page_invert">Invert dual page split placement</string>
     <string name="pref_dual_page_invert_summary">If the placement of the dual page split doesn\'t match reading direction</string>
+    <string name="pref_long_strip_split">Long Strip Split (Alpha)</string>
+    <string name="pref_long_strip_split_summary">Impore reader performance by splitting long strips</string>
```
Comment (arkon) [inline app/src/main/res/values/strings.xml:300]: Can just reuse `split_tall_images_summary`

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -296,6 +296,8 @@
     <string name="pref_dual_page_split">Dual page split</string>
     <string name="pref_dual_page_invert">Invert dual page split placement</string>
     <string name="pref_dual_page_invert_summary">If the placement of the dual page split doesn\'t match reading direction</string>
+    <string name="pref_long_strip_split">Long Strip Split (Alpha)</string>
```
Comment (arkon) [inline app/src/main/res/values/strings.xml:299]:
```suggestion
    <string name="pref_long_strip_split">Split tall images (Alpha)</string>
```

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -296,8 +296,8 @@
     <string name="pref_dual_page_split">Dual page split</string>
     <string name="pref_dual_page_invert">Invert dual page split placement</string>
     <string name="pref_dual_page_invert_summary">If the placement of the dual page split doesn\'t match reading direction</string>
-    <string name="pref_long_strip_split">Long Strip Split (Alpha)</string>
-    <string name="pref_long_strip_split_summary">Impore reader performance by splitting long strips</string>
+    <string name="pref_long_strip_split">Split tall images (Alpha)</string>
+    <string name="pref_long_strip_split_summary">@string/split_tall_images_summary</string>
```
Comment (arkon) [inline app/src/main/res/values/strings.xml:300]: Just use the string directly.

**Pull request merged**